mitiq/lre/inference/multivariate_richardson.py

```diff
@@ -0,0 +1,199 @@
+        for var in sorted(counts.keys()):
+            count = counts[var]
+            if count > 1:
+                monomial_parts.append(f"{var}**{count}")
+            else:
+                monomial_parts.append(var)
+        monomial = "*".join(monomial_parts)
+        # Handle the case where degree is 0
+        # the tuple in variable_combinations is empty
+        if not monomial:
+            monomial = "1"
+        monomial_basis.append(monomial)
+    # "1" should be the first monomial (degree = 0).
+    # max degree should be the last term
+    return monomial_basis[::-1]
+
+
+def sample_matrix(
+    input_circuit: Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: Optional[int] = None,
+) -> NDArray[Any]:
+    r"""
+    Defines the sample matrix required for multivariate extrapolation as
+    defined in :cite:`Russo_2024_LRE`.
+
+    Args:
+        input_circuit: Circuit to be scaled.
+        degree: Degree of the multivariate polynomial.
+        fold_multiplier: Scaling gap required by unitary folding.
+        num_chunks: Number of desired approximately equal chunks. When the
+            number of chunks is the same as the layers in the input circuit,
+            the input circuit is unchanged.
+
+    Returns:
+        Matrix of the evaluated monomial basis terms from the scale factor
+            vectors.
+
+    Raises:
+        ValueError:
+            When the degree for the multinomial is not greater than or
+                equal to 1; when the fold multiplier to scale the circuit is
+                greater than/equal to 1; when the number of chunks for a
+                large circuit is 0 or when the number of chunks in a circuit is
+                greater than the number of layers in the input circuit.
+
+    """
+    if degree < 1:
+        raise ValueError(
+            "Multinomial degree must be greater than or equal to 1."
+        )
+    if fold_multiplier < 1:
+        raise ValueError("Fold multiplier must be greater than or equal to 1.")
+
+    scale_factor_vectors = _get_scale_factor_vectors(
+        input_circuit, degree, fold_multiplier, num_chunks
+    )
+    num_layers = len(scale_factor_vectors[0])
+
+    monomial_terms = full_monomial_basis(num_layers, degree)
+    if len(monomial_terms) != len(scale_factor_vectors):
+        raise ValueError("Sample matrix will not be a square matrix.")
+    sample_matrix = np.zeros((len(monomial_terms), len(monomial_terms)))
+
+    # Evaluate the monomial terms using the values in the scale factor vectors
+    # and insert in the sample matrix
+    # each row is specific to each scale factor vector
+    # each column is a term in the monomial basis
+    for i, point in enumerate(scale_factor_vectors):
+        for j, monomial in enumerate(monomial_terms):
+            var_mapping = {f"λ_{k+1}": point[k] for k in range(num_layers)}
+            sample_matrix[i, j] = eval(monomial, {}, var_mapping)
```

Comment:
Using `eval()` can be [dangerous](https://stackoverflow.com/questions/1832940/why-is-using-eval-a-bad-practice). 

Consider writing and using something like this as an alternative: 

```py
def evaluate_monomial(monomial: str, var_mapping: dict[str, float]) -> float:
    """Safely evaluate a monomial string."""
    if monomial == "1":
        return 1.0
    terms = monomial.split("*")
    result = 1.0
    for term in terms:
        if "**" in term:
            var, power = term.split("**")
            result *= var_mapping[var] ** int(power)
        else:
            result *= var_mapping[term]
    return result
    ```

Reply:
I think we should generally try to avoid storing the monomials as strings as this requires us to do lots of string manipulation. Why use strings when we can live in number land the whole time?

To this end, what if we store the monomials as tuples of exponents instead of strings? E.g. $x_0^0x_1^3x_2^1x_3^2$ gets stored as $(0, 3, 1, 2)$ whereupon we can then use an evaluate function that looks a little simpler:[^1]
```py
def evaluate_monomial(exponents: tuple[int, ...], values: list[float]) -> float:
    result = 1
    for exp, value in zip(exponents, values):
        result *= value ** exp
    return result
```

WDYT? If we went this route, the `full_monomial_basis` function can be simplified quite a bit to something like
```py
def full_monomial_basis(n: int, d: int) -> List[Tuple[int, ...]]:
    return [exps for exps in itertools.product(range(d + 1), repeat=n) if sum(exps) <= d]
```


[^1]: code not guaranteed to work :)

Reply:
> To this end, what if we store the monomials as tuples of exponents instead of strings?

I like this idea!